app/src/main/java/org/simple/clinic/security/pin/PinEntryCardController.kt

```diff
@@ -48,35 +55,73 @@ class PinEntryCardController @Inject constructor(
+
+          val recordAttempts = cachedPinValidation
+              .switchMap {
+                when (it) {
+                  SAME -> bruteForceProtection.recordSuccessfulAuthentication()
+                  DIFFERENT -> bruteForceProtection.incrementFailedAttempt()
+                }.andThen(Observable.empty<UiChange>())
+              }
 
           val validationResultUiChange = cachedPinValidation
               .map {
-                { ui: Ui ->
-                  when (it) {
-                    SAME -> ui.dispatchAuthenticatedCallback()
-                    DIFFERENT -> {
-                      ui.showIncorrectPinError()
-                      ui.clearPin()
-                    }
-                  }
+                when (it) {
+                  SAME -> { ui: Ui -> ui.dispatchAuthenticatedCallback() }
+                  DIFFERENT -> { ui: Ui -> ui.clearPin() }
                 }
               }
-              .toObservable()
 
-          val progressUiChanges = cachedPinValidation
-              .filter { it == ComparisonResult.DIFFERENT }
-              .map { { ui: Ui -> ui.moveToState(State.PinEntry) } }
-              .toObservable()
-              .startWith { ui: Ui -> ui.moveToState(State.Progress) }
+          Observable.mergeArray(progressUiChanges, recordAttempts, validationResultUiChange)
+        }
+  }
 
-          Observable.mergeArray(progressUiChanges, validationResultUiChange)
+  private fun blockWhenAuthenticationLimitIsReached(events: Observable<UiEvent>): Observable<UiChange> {
+    return events.ofType<PinEntryViewCreated>()
+        .flatMap { bruteForceProtection.protectedStateChanges() }
+        .switchMap { state ->
+          when (state) {
+            ProtectedState.Allowed -> {
+              Observable.just({ ui: Ui -> ui.moveToState(State.PinEntry) })
+            }
+
+            is ProtectedState.Blocked -> {
+              Observable.interval(1L, TimeUnit.SECONDS)
+                  .startWith(0L)  // initial item.
+                  .map {
+                    val formattedTimeRemaining = formatTimeRemainingTill(state.blockedTill)
+                    return@map { ui: Ui ->
+                      ui.moveToState(State.BruteForceLocked(formattedTimeRemaining))
+                    }
+                  }
+            }
+          }
         }
   }
 
+  private fun formatTimeRemainingTill(futureTime: Instant): String {
+    val secondsRemaining = futureTime.epochSecond - Instant.now(clock).epochSecond
+    val secondsPerHour = Duration.ofHours(1).seconds
+    val secondsPerMinute = Duration.ofMinutes(1).seconds
+
+    val minutes = (secondsRemaining % secondsPerHour / secondsPerMinute).toString()
+    val seconds = (secondsRemaining % secondsPerMinute).toString()
+
+    val minutesWithPadding = minutes.padStart(2, padChar = '0')
+    val secondsWithPadding = seconds.padStart(2, padChar = '0')
+
+    return "$minutesWithPadding:$secondsWithPadding"
```

Comment:
What do you think of this, @vinaysshenoy? I'm not using any date formatter so that I can easily check the return value from unit tests.

Reply:
> I think this should just emit Duration or a value type that just holds hours and seconds which is tested in the controller.

I considered doing this, but generating formatted timer from a `Duration` object requires Maths and I am not comfortable keeping date Maths in the UI. 

I could… create a `TimerDuration` object and pass it to the UI. What do you think?